### DISCUSSION TITLE
[General] Created a PR for SignatureBased Paymaster in paymaster-examples since few people are struggling here. Hope it helps :)

### DISCUSSION THREAD

**Discussion by hoshiyari420** [zkSync-Community-Hub/zksync-developers#405] [Show and Tell]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

https://github.com/matter-labs/paymaster-examples/pull/22

I am sharing the pull request here so that few folks who are struggling with `innerInputs` encoding might find it useful. 
The PR is yet to be reviewed and merged, hence work in progress. Be cautious before using it in production as the contracts are examples and not designed for production.

I hope the unit tests and the `createSignatureData()` function help solving errors regarding signature validation issues. 

#### Functionality:
- This paymaster contract shall validate based on the signature provided through `innerInputs` whether or not the user's transaction shall be allowed to proceed or not. 
- The signer shall have authority to ensure signature is only valid upto particular timestamp. 
- The signature shall be EIP712 compliance which is signed on values like `(user address, last timestamp, nonces)`. 
- Users are required to provide certain inputs such as signature and last timestamp as `innerInputs` in paymaster params. 
- The paymaster will then check signature based on `transaction.from`, `block.timestamp` and `nonces` to ensure only valid users are allowed to use paymaster. 

#### Security Considerations:
- Paymaster should use nonce for every user in validating signatures which increments on every successful transactions.
- Paymaster should check expiry timestamp of signatures. 
- Paymaster should validate signatures based on `transaction.from` to ensure correct user has sent transaction. 
- Paymaster should revert if recovered signer does not match or signature expires. 
- Owner can update the paymaster signer.
- Owner should be able to cancel any user's nonce. 
- EIP712 compliance signatures support. 
- Use of battle-tested standard libraries for ECDSA signature verification(Openzeppelin). 

### What is the specific issue or error you're encountering?

Encoding issues

### Can you share the error messages or logs you're receiving, if any?

Invalid signature validation

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

Openzeppelin

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

N/A

### What steps have you already taken to try to resolve the issue?

N/A

### Repo Link (Optional)

https://github.com/matter-labs/paymaster-examples/pull/22

### Additional Details

- Not yet merged. Please be cautious.

**Comment by uF4No** [comment]
hey @hoshiyari420 thanks for taking the time to work on this and sharing it 🤘 We'll review the PR and give you some feedback or merge it 🤙

Btw, I've moved it to the "Show & Tell" category.

**Comment by chudny21** [comment]
LFG

**Comment by Blueeloop** [comment]
Niceee.

**Comment by Rainbowteeth** [comment]
This is awesome. This  opens up some interesting usecases.

**Reply by hoshiyari420** [reply]
Yes, for eg: pay for gas for all of your connected/related addresses through this paymaster without "allow-list type" transaction.
